Use SystemLimit instead of settings for device rate limiting

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-16355
https://dimagi.atlassian.net/browse/SAAS-16561

I recently added a `SystemLimit` class that makes it much easier to manage limits, and moving this limit over from settings to the SystemLimit in this PR. There are a few other changes that go along with this refactor, like moving the setting to enable the device rate limiter over to a feature flag, and a slight bug fix (sort of) related to the redis key not being scoped to a domain.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
DEVICE_RATE_LIMITER

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No, I'll test myself.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Add for_key classmethod on SystemLimit

gherceg committed:
Ensure redis key is unique per domain

Realistically I don't think this is a likely scenario in our current
platform since a mobile worker is scoped to a domain, but noticed this
vulnerability in the test and decided it was easy enough to protect
against.

gherceg committed:
Replace settings limits with SystemLimit

This commit did a lot. In addition to replacing the limit in settings
with SystemLimit, I removed the feature flag for an increased limit
since SystemLimit solves that, and moved the setting to enable the
device rate limiter to a feature flag for ease of use. I also updated
metrics to differentiate between an actual rate limited attempt vs a
would-be rate limited.

gherceg commented:
@coderabbitai review

orangejenny commented on corehq/project_limits/tests/test_system_limit.py:
```diff
@@ -0,0 +1,18 @@
+    def test_for_key_returns_domain_specific_limit(self):
+        SystemLimit.objects.create(key="general_limit", limit=10)
+        SystemLimit.objects.create(key="general_limit", limit=20, domain="specific")
+        self.assertEqual(SystemLimit.for_key("general_limit", domain="specific"), 20)
```
I'd add an assert that calling `for_key` without a domain still returns 10.

gherceg replied:
[8bbb387](https://github.com/dimagi/commcare-hq/pull/35781/commits/8bbb387a888396cd3858e34bb4ba5c040ff48a4a)

millerdev commented on corehq/project_limits/models.py:
```diff
@@ -97,3 +97,15 @@ class Meta:
+            domain_filter |= models.Q(domain=domain)
+        filters = models.Q(key=key) & domain_filter
+        limit = SystemLimit.objects.filter(filters).order_by("-domain").values_list("limit", flat=True).first()
+        return limit or None
```
What are the possible values of `limit` that evaluate to false?

gherceg replied:
Ah great catch. It is a positive integer field, so 0 is the only possible falsey value in addition to None. Updated to fix that [8bbb387](https://github.com/dimagi/commcare-hq/pull/35781/commits/8bbb387a888396cd3858e34bb4ba5c040ff48a4a)

millerdev commented on corehq/project_limits/tests/test_system_limit.py:
```diff
@@ -0,0 +1,18 @@
+
+    def test_for_key_returns_general_limit(self):
+        SystemLimit.objects.create(key="general_limit", limit=10)  # domain defaults to blank
+        self.assertEqual(SystemLimit.for_key("general_limit"), 10)
```
Maybe it's overkill, but would it make sense to add a test similar to this one except with
```py
self.assertEqual(SystemLimit.for_key("general_limit", domain="no_match"), 10)
```
Or maybe just add that extra assertion in this test?

gherceg replied:
[8bbb387](https://github.com/dimagi/commcare-hq/pull/35781/commits/8bbb387a888396cd3858e34bb4ba5c040ff48a4a)

millerdev commented on corehq/apps/users/device_rate_limiter.py:
```diff
@@ -63,10 +64,14 @@ def rate_limit_device(self, domain, user_id, device_id):
-            'commcare.devices_per_user.rate_limit_exceeded', tags={'domain': domain, 'user_id': user_id}
-        )
-        return settings.ENABLE_DEVICE_RATE_LIMITER
+        if toggles.DEVICE_RATE_LIMITER.enabled(domain):
+            metrics_counter('commcare.devices_per_user.rate_limited', tags={'domain': domain, 'user_id': user_id})
+            return True
+        else:
+            metrics_counter(
+                'commcare.devices_per_user.rate_limited.test', tags={'domain': domain, 'user_id': user_id}
+            )
+            return False
```
What do you think of putting the toggle value in a tag?
```suggestion
        enabled = toggles.DEVICE_RATE_LIMITER.enabled(domain):
        metrics_counter(
            'commcare.devices_per_user.rate_limited',
            tags={'domain': domain, 'user_id': user_id, 'enabled': str(enabled)},
        )
        return enabled
```
That would allow easy grouping in DD, but also combine the two metrics by default, which is admittedly maybe not what you want.

gherceg replied:
Yeah I'm fine with this. Should still be able to do the separation in datadog regardless of a different metric vs tags. [8bbb387](https://github.com/dimagi/commcare-hq/pull/35781/commits/8bbb387a888396cd3858e34bb4ba5c040ff48a4a)

millerdev commented on corehq/apps/users/tests/test_device_rate_limiter.py:
```diff
@@ -1,77 +1,100 @@
+        SystemLimit.objects.create(key=DEVICE_LIMIT_PER_USER_KEY, limit=1)
         self.addCleanup(clear_redis)
 
+    @flag_enabled("DEVICE_RATE_LIMITER")
```
Can this decorator not be applied to the class? If not, wonder how hard it would be to change it to work at the class level?

gherceg replied:
it works at the class level, but then `@flag_disabled` didn't work for the one test where I wanted to use it since the patch had already been applied for the `@flag_enabled` decorator. I briefly looked to see if it would be an easy fix but came up empty handed and decided to just do this instead.

millerdev replied:
:slightly_frowning_face:

esoergel replied:
I added these decorators originally and I vaguely recall also running into that shortcoming and not seeing an easy fix.  Would love to see that addressed if anyone can figure it out.

gherceg replied:
Added it as a hackathon idea for tomorrow 🤷🏻

gherceg replied:
Shoutout @millerdev https://github.com/dimagi/commcare-hq/pull/35818

gherceg replied:
2d5aea65bd88ca69793c69f27ce2f12afc3334e6

orangejenny commented on corehq/apps/users/tests/test_device_rate_limiter.py:
```diff
@@ -1,77 +1,100 @@
         self.assertFalse(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id'))
 
-    @override_settings(DEVICE_LIMIT_PER_USER=2)
+    @flag_enabled("DEVICE_RATE_LIMITER")
```
Can you do this on the class itself?

gherceg replied:
it works at the class level, but then @flag_disabled didn't work for the one test where I wanted to use it since the patch had already been applied for the @flag_enabled decorator. I briefly looked to see if it would be an easy fix but came up empty handed and decided to just do this instead.

orangejenny replied:
that's annoying

gherceg replied:
2d5aea65bd88ca69793c69f27ce2f12afc3334e6 (thank you @millerdev)

orangejenny commented on corehq/toggles/__init__.py:
```diff
@@ -2995,13 +2995,13 @@ def domain_has_privilege_from_toggle(privilege_slug, domain):
     namespaces=[NAMESPACE_DOMAIN],
 )
 
-INCREASE_DEVICE_LIMIT_PER_USER = StaticToggle(
```
Nobody has this turned on, right?

gherceg replied:
Correct. And if they do, they shouldn't 😅

gherceg committed:
Add quickcache to SystemLimit.for_key

gherceg committed:
Address PR feedback

- add more assertions in SystemLimit.for_key tests
- handle case where limit is 0
- use tag to determine if rate limited metric was enforced or not

gherceg commented:
Expecting tests to fail due to not clearing redis. Offlining with @millerdev about what the proper solution is.

gherceg committed:
Fix bug with clearing cache properly

millerdev commented on corehq/project_limits/models.py:
```diff
@@ -97,3 +99,21 @@ class Meta:
+        if domain:
+            domain_filter |= models.Q(domain=domain)
+        filters = models.Q(key=key) & domain_filter
+        limit = SystemLimit.objects.filter(filters).order_by("-domain").values_list("limit", flat=True).first()
+        return None if limit is None else limit  # limit could be 0
```
nit: How is this different from `return limit`?
```suggestion
        return SystemLimit.objects.filter(filters).order_by("-domain").values_list("limit", flat=True).first()
```

gherceg replied:
[11c4e63](https://github.com/dimagi/commcare-hq/pull/35781/commits/11c4e63bb2fee263f47da3ec7726af544d2be5f2)

gherceg committed:
Just return the limit

gherceg committed:
Fix gaps in caching logic

- For testing, skip caching for_key as it creates too many problems
- Make sure to clear the cache when a SystemLimit is deleted too

gherceg committed:
Avoid clearing redis in device rate limiter tests

Clearing all of redis masked issues with the caching logic in the
SystemLimit.for_key which isn't good. The right approach is to skip
caching when running in a test.

gherceg commented:
Alright no more commits! Sorry for the churn.

gherceg committed:
Cache more wisely

This doesn't actually matter since the data being saved in redis is very
small, but this change caches more wisely. Previously each key+domain
combo would have a cached key, even though in most cases, a domain will
not have a unique value from the global value. This change only stores a
key+domain value in cache when necessary.

gherceg commented:
I lied. Made a slight change to the caching strategy for SystemLimit that came up during the dev block. No more changes starting _now_.

gherceg commented on corehq/project_limits/models.py:
```diff
@@ -97,3 +100,41 @@ class Meta:
+
+    def delete(self, *args, **kwargs):
+        super().delete(*args, **kwargs)
+        self._get_global_limit.clear(self.__class__, self.key)
+        self._get_domain_specific_limit.clear(self.__class__, self.key, self.domain)
```
I could be smarter about only clearing the cache for the specific methods when relevant (if `self.domain` is a blank string vs something specific)

gherceg replied:
https://github.com/dimagi/commcare-hq/pull/35781/commits/ce7b0102aeada48e9af3adf4ea55d0dcd231604d

gherceg replied:
...hmm but what someone modifies an existing limit to be scoped to domain (or vice versa). That would not be an ideal usage, but would result in us not clearing the correct cache with this updated logic...

gherceg committed:
Only clear relevant cache

We know which cache to clear depending on the domain field of the object
being saved/deleted.

millerdev commented on corehq/project_limits/models.py:
```diff
@@ -103,23 +103,38 @@ def __str__(self):
-            domain_filter |= models.Q(domain=domain)
-        filters = models.Q(key=key) & domain_filter
-        return SystemLimit.objects.filter(filters).order_by("-domain").values_list("limit", flat=True).first()
+        domain_limit = cls._get_domain_specific_limit(key, domain) if domain else None
```
That might complicate cache clearing though since the global limit would be cached per domain, and those would not be cleared when the global limit is changed.

I'm curious how much more performant this is with caching applied? The SQL query should be very fast, possibly not much slower than hitting redis. Would be interesting to measure.

gherceg replied:
Yeah I'm up for measuring the difference, though am inclined to start by just hitting the cache and do that in a followup PR.

gherceg committed:
Revert "Only clear relevant cache"

This reverts commit ce7b0102aeada48e9af3adf4ea55d0dcd231604d.

It is possible to modify an existing limit and add a domain, in which
case we would be clearing the wrong cache, so we need to revert.

gherceg committed:
Manually cache domain specific limit

millerdev commented on corehq/project_limits/models.py:
```diff
@@ -97,3 +101,52 @@ class Meta:
+        return cls._get_global_limit(key)
+
+    @staticmethod
+    @quickcache(['key'], timeout=7 * 24 * 60 * 60, skip_arg=lambda *args, **kwargs: settings.UNIT_TESTING)
```
Forgot to mention this earlier, but test code mixed with production code is not ideal. Would it be possible to use a patch to disable this quickcache for the tests that need to disable it?

Might be able to do this with a package- or module-scoped pytest fixture similar to

https://github.com/dimagi/commcare-hq/blob/a95068b2a6cbe30d6496526e53c198d7a873ed8f/corehq/apps/experiments/tests/test_experiment.py#L194-L200

It will probably be a more complicated patch than those, depending on how surgical you need to be to disable quickcache.

gherceg replied:
> Would it be possible to use a patch to disable this quickcache for the tests that need to disable it?

What I don't like about this is that I imagine tests (like the test_device_rate_limiter) that use SystemLimit under the hood would then need to know to disable cache specifically for that SystemLimit method. To your point, maybe just hitting the database is better since it avoids caching entirely, but I think if we do use cache for this method, the best way to reduce friction when writing tests for classes that rely on SystemLimit is to ensure the cache is disabled by default.

gherceg replied:
On the flip side, I would like a way to _enable_ the cache so that I could test that it actually clears how I expect it to in `test_system_limit`, but maybe that is overkill.

gherceg replied:
We offlined, but came to the agreement that using a local cache would ensure that even if this class sees high usage and thus high load, hitting a local cache would be most ideal to avoid network requests. We also agreed that removing the complexity of clearing the cache with the tradeoff of waiting up to 5 minutes for a limit change to take effect was worth it. 

d6a6defcfc4e24b3de37e6c9602af10519fbff37

gherceg committed:
Merge branch 'master' into gh/device-limit/use-system-limit

gherceg committed:
Local cache with 5 min timeout instead of clearing

In the worst case, it'll take 5 minutes for a new limit to take effect,
but we don't have to worry about clearing cache.

This also uses local memory instead of redis, and because the underlying
db query isn't too expensive, it should be more efficient to rely on a
shorter cache life in local memory than a longer cache life in redis.

gherceg committed:
Set flag_enabled at class level

millerdev commented on corehq/apps/users/tests/test_device_rate_limiter.py:
```diff
@@ -1,27 +1,30 @@
+
+    def tearDown(self):
+        for key in device_rate_limiter.client.scan_iter(f"{REDIS_KEY_PREFIX}*"):
+            device_rate_limiter.client.delete(key.decode('utf-8'))
```
Is this still needed since the global cache patch has been added?

gherceg replied:
Yeah the global cache patch is for SystemLimit, which is just used by the device rate limiter, but the rate limiter still stores its relevant info in redis and needs to clear it after test runs.

millerdev commented on corehq/project_limits/tests/test_system_limit.py:
```diff
@@ -23,7 +23,9 @@ def test_for_key_returns_domain_specific_limit(self):
 
-    def test_cache_is_cleared(self):
+    def test_patching_cache_works(self):
+        # if the logic in ../patch_cache.py fails, this test should catch that
```
nit: `patch_cache.py` is in the same directory as this test module, correct?
```suggestion
        # if the logic in patch_cache.py fails, this test should catch that
```

gherceg replied:
a091831ffba8c94800afe0b2a1be93bd7fd41171

gherceg committed:
Address feedback

gherceg commented on corehq/project_limits/tests/patch_cache.py:
```diff
@@ -0,0 +1,14 @@
+def patch_systemlimit_cache():
+    assert settings.UNIT_TESTING
+    SystemLimit._get_global_limit = SystemLimit._get_global_limit.__wrapped__
+    # Use __enter__ and __exit__ to start/stop so patch.stopall() does not stop it.
```
a091831ffba8c94800afe0b2a1be93bd7fd41171

gherceg committed:
Revert to simpler cache

This references settings.UNIT_TESTING in
production code to skip caching, but this is a less fragile approach
than the solution I had to remove the quickcache decorator by blindly
unwrapping the function, and is used throughout our code.

This uses redis in addition to local memory caching, but that is
how it is typically done in HQ and has worked well for us, and results
in a simpler decorat that is easier for future readers to understand.

This stores redundant data in the cache, but avoids hitting the db again
which is what we want.

gherceg committed:
Specify namespace when checking toggle

millerdev commented on corehq/project_limits/models.py:
```diff
@@ -97,3 +102,39 @@ class Meta:
+        return cls._get_global_limit(key)
+
+    @staticmethod
+    @quickcache(['key'], timeout=FIVE_MINUTES, skip_arg=lambda *args: settings.UNIT_TESTING)
```
Is it fine that this caching is scoped to request/celery task? (That's the default behavior of quickcache without a custom `session_function`.)

Aside: I wonder how quickcache works in pillow processes? I assume a it's the same as a single celery task.

gherceg replied:
> Is it fine that this caching is scoped to request/celery task? (That's the default behavior of quickcache without a custom session_function.)

Local caching is scoped to request/task right? As my other comment outlines, I think hitting redis frequently to start is fine, and I can monitor to see if additional changes are needed.

millerdev reviewed (approved):
Approving to allow this work to move forward, but I have to admit, I do not like the [revert commit](https://github.com/dimagi/commcare-hq/pull/35781/commits/dc1cc442ffe3ac10ee1426593aabbc41a8075717). It made this review difficult because it was unclear whether comments I had made earlier are still addressed. I felt like I needed to start over and review the entire change set from the beginning.

From https://github.com/dimagi/commcare-hq/pull/35781/commits/de45dc7da9a1584f933808942303826474b48f49#r1972225590

> We ... came to the agreement that using a local cache would ensure that even if this class sees high usage and thus high load, hitting a local cache would be most ideal to avoid network requests. We also agreed that removing the complexity of clearing the cache with the tradeoff of waiting up to 5 minutes for a limit change to take effect was worth it.

Do you no longer agree to those things?

Was that revert commit motivated by the difficulty of resolving https://github.com/dimagi/commcare-hq/pull/35781#discussion_r1972235005? I think we could find a solution.

Looking forward to the follow-up measurements to compare caching vs not.

gherceg commented:
> > We ... came to the agreement that using a local cache would ensure that even if this class sees high usage and thus high load, hitting a local cache would be most ideal to avoid network requests. We also agreed that removing the complexity of clearing the cache with the tradeoff of waiting up to 5 minutes for a limit change to take effect was worth it.

> Do you no longer agree to those things?

Hitting a local cache is certainly more performant, and if it is clear that this change results in an increase in [form submission timings](https://app.datadoghq.com/dashboard/g9s-pw6-tpg/hq-vitals?fromUser=false&fullscreen_end_ts=1741708595952&fullscreen_paused=false&fullscreen_refresh_mode=sliding&fullscreen_section=overview&fullscreen_start_ts=1741103795952&fullscreen_widget=200981468&refresh_mode=sliding&from_ts=1741103746436&to_ts=1741708546436&live=true), I'll make the necessary changes to support hitting a local cache instead. However given how much we hit redis on any given request, I'm not terribly concerned about the performance impact of using the redis cache instead to start.

Clearing the cache in this instance seems straightforward. Apart from the one edge case identified in tests (the cache not being cleared if a transaction is rolled back), I'm not concerned about a stale cache, and we decided that is too much of an edge case to accommodate.

> Was that revert commit motivated by the difficulty of resolving https://github.com/dimagi/commcare-hq/pull/35781#discussion_r1972235005? I think we could find a solution.

Yes and no. That seems like a solvable problem, but the deeper I got into these changes the more I felt like it was simpler to go with the status quo, even if the status quo isn't best practice. While patching the cached methods in test setup removed test code from production, it also made discoverability of that patch very hard.

I think we both agree that we need a more robust and safer way to disable quickcache in tests by default, and explicitly enable it when desired, and at some point felt like I was trying to do that in this PR rather than separately.

> Looking forward to the follow-up measurements to compare caching vs not.

Created [this ticket](https://dimagi.atlassian.net/browse/SAAS-16778)
